### PR TITLE
restrict C# nuget unit tests to local package feeds

### PR DESCRIPTION
**Pull request by brettfo**
Comments have been added inline with the code where the most attention should be paid.

In short, if a unit test is written as it was before, the behavior will remain the same; network access to `api.nuget.org` will be allowed.  _However_, as soon as the `packages` parameter is specified, a test-only `NuGet.Config` is written to the test directory which disallows **all** network access and only allows the packages created in a special directory to be queried.  Since we don't use any custom NuGet API handling, this is perfectly fine as we're letting the "real" NuGet handle its sources however it sees fit (which in this case is all from a local feed.)

The vast majority of this PR was changing things like `Newtonsoft.Json` to `Some.Package` and can be ignored.  In some tests the specific package dependencies that made the test case interesting have been explicitly laid out which should help with readability quite a bit.

### CODE REVIEW

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs [hunk 1]
```diff
@@ -201,6 +223,42 @@ public abstract class UpdateWorkerTestBase : TestBase
         AssertContainsFiles(expectedResult, actualResult);
     }
 
+    public static async Task MockNuGetPackagesInDirectory(MockNuGetPackage[]? packages, string temporaryDirectory)
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:226]: This is where the fake NuGet packages are written to a local feed and an appropriate `NuGet.Config` is created.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/MockNuGetPackage.cs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        private static readonly XmlWriterSettings WriterSettings = new()
+        {
+            Encoding = Encoding.UTF8,
+            Indent = true,
+        };
+
+        private XDocument? _nuspec;
+        private Stream? _stream;
+
+        public void WriteToDirectory(string localPackageSourcePath)
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/MockNuGetPackage.cs:33]: This is the interesting part about the fake NuGet packages; a directory is populated with some specific files (comments inline) that allow the directory to be used as a NuGet package source.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs (hunk 1 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:226]: If no `packages` parameter is specified, no local feed and no `NuGet.Config` file will be created, which will still allow tests to hit the network if they're still required.  As soon as a test sets `packages: [...]` then all network access it cut off.  I did this to allow for the most flexibility, but also as a way to ensure that a test that specifies `packages` won't accidentally pull something from `api.nuget.org`.

**Review by brbayes-msft — APPROVED** [review]

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/PathHelper.cs [hunk 3]
```diff
@@ -111,4 +111,22 @@ public static void CopyDirectory(string sourceDirectory, string destinationDirec
             CopyDirectory(subDir.FullName, newDestinationDir);
         }
     }
+
+    public static bool IsSubdirectoryOf(string parentDirectory, string childDirectory)
+    {
+        var parentDirInfo = new DirectoryInfo(parentDirectory);
+        var childDirInfo = new DirectoryInfo(childDirectory);
+
+        while (childDirInfo.Parent is not null)
```
Comment (JoeRobich) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/PathHelper.cs:120]: I like this more thorough implementation, but would `childDirInfo.FullName.StartsWith(parentDirInfo.FullName);` work? This is me assuming DirectoryInfo normalized the paths.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs [hunk 4]
```diff
@@ -201,6 +223,42 @@ public abstract class UpdateWorkerTestBase : TestBase
...
+            MockNuGetPackage[] allPackages = packages.Concat(MockNuGetPackage.CommonPackages).ToArray();
+
+            // write all packages to disk
+            foreach (MockNuGetPackage package in allPackages)
+            {
+                package.WriteToDirectory(localFeedPath);
+            }
+
+            // override various nuget locations
+            foreach (var envName in new[] { "NUGET_PACKAGES", "NUGET_HTTP_CACHE_PATH", "NUGET_SCRATCH", "NUGET_PLUGINS_CACHE_PATH" })
```
Comment (JoeRobich) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:241]: Do these values need to be reverted or is the expectation that every test will end up mocking their packages?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs (hunk 1 above)
Comment (JoeRobich) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:226]: I wonder if this method should be extracted somewhere more central, since it isn't Updater specific.

**Review by JoeRobich — APPROVED** [review]
This change makes me so happy!

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/PathHelper.cs (hunk 3 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/PathHelper.cs:120]: `.StartsWith()` doesn't work for things like `C:\temp` and `C:\temporary-documents\file.txt`.  In this case `.StartsWith` will return `true`, but it's not under the same path.  I considered ensuring a trailing directory separator, but that ends up with a bunch of edge cases that I didn't like.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs (hunk 4 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:241]: They're re-set in every test and the only negative effect I could think of is if a test _doesn't_ use the fake NuGet packages and actually wants to hit `api.nuget.org`, it's possible that these values will cause that test to break, but since the goal of this is to isolate the tests, that kind of test breaking is probably a good thing.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs (hunk 1 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Update/UpdateWorkerTestBase.cs:226]: I made it `public` so I could use it elsewhere (in fact, the `discover` tests use it), I just didn't have a good place to put it and I wanted to get this huge PR off of my box :)
